### PR TITLE
Add monitoring and alerting for realm capacity

### LINKED ISSUE #572: Set up monitoring and alerting when realms are at certain capacity levels
Set up monitoring and alerting when realms are at certain capacity levels (e.g. 90%). These will have to go to the server operators since we don't have a way to send email to realm admins.

There are already stats - we just need the monitoring and alerts.

/cc @icco @yegle - do one of you have capacity to work on this?

### PR DESCRIPTION
**Pull request by femnad**
Fixes #572

## Proposed Changes

* Record metric for realm capacity on issuing a verification token
* Add dashboard and alerting for monitoring realm capacity

**Release Note**

```release-note
Record and monitor realm verification token capacity
```

### CODE REVIEW

File: pkg/controller/issueapi/issue.go
Comment (sethvargo) [inline pkg/controller/issueapi/issue.go:258]: Use:

```go
if err := c.recordCapacity(ctx, realm); err != nil {
...
```

to avoid shadowing err

File: pkg/controller/issueapi/issue.go
Comment (sethvargo) [inline pkg/controller/issueapi/issue.go:261]: Is this the right stat?

File: pkg/controller/issueapi/issue.go
Comment (sethvargo) [inline pkg/controller/issueapi/issue.go:262]: Are you sure we want to fail if we don't update this stat?

File: pkg/controller/issueapi/issue.go
Comment (sethvargo) [inline pkg/controller/issueapi/issue.go:326]: I'm not sure I understand why this is 1 - %?

**Review by sethvargo — CHANGES_REQUESTED** [review]
Hmm I was more thinking we'd generate a metric/alert when a realm reached their capacity (or when the capacity was less than some small number like "10". The result from `Take` should include all the info you need without making the extra db calls.

File: pkg/controller/issueapi/issue.go
Comment (femnad) [reply pkg/controller/issueapi/issue.go]: It felt wrong to alert above a capacity of 90% but maybe the wording on the alert should be "capacity utilization" so changed to that.

File: pkg/controller/issueapi/issue.go
Comment (femnad) [reply pkg/controller/issueapi/issue.go]: In my mind associated that error with the whole process, but now it's not fatal.

File: pkg/controller/issueapi/issue.go
Comment (femnad) [reply pkg/controller/issueapi/issue.go]: Capacity recording does not return an error anymore.

**Comment by femnad** [comment]
> Hmm I was more thinking we'd generate a metric/alert when a realm reached their capacity (or when the capacity was less than some small number like "10". The result from `Take` should include all the info you need without making the extra db calls.

Seeing the return of `Take` now I see that would have been more straightforward. Change to comparing the remaining to 10 (constant) and have a two-state alert based on that?

File: pkg/controller/issueapi/issue.go
Comment (sethvargo) [inline pkg/controller/issueapi/issue.go:258]: Yea, so we know their total configured limit and the number remaining. I'm gonna defer to @icco on whether we want to alert on a fixed val (e.g. 10) or a ratio here (e.g. < 10% quota remaining).

**Review by sethvargo — COMMENTED** [review]
/assign @icco

File: pkg/controller/issueapi/issue.go
Comment (icco) [reply pkg/controller/issueapi/issue.go:258]: We wanna alert on percentage, but the metrics should be the actual numbers. So record remaining and total here, and in the alert calculate the percentage

Comment (femnad) [reply pkg/controller/issueapi/issue.go:258]: Can we have alerting based on multiple metrics? From what I can tell Stackdriver only seemed to consider a single metric, am I wrong?

**Review by icco — COMMENTED** [review]
.

File: pkg/controller/issueapi/issue.go
Comment (icco) [reply pkg/controller/issueapi/issue.go:258]: Uhm, my understanding was that we can do math now with Query Notation, but I have not tried

Comment (femnad) [reply pkg/controller/issueapi/issue.go:258]: Yeah, I can do something like the following to calculate memory utilization in metrics explorer via MQL:

```
{ fetch 'compute.googleapis.com/instance/memory/balloon/ram_used'
; fetch 'compute.googleapis.com/instance/memory/balloon/ram_size' }
| join
| div
```
however, couldn't make something similar work via Terraform alert policy filter, it doesn't seem possible to select multiple metrics by `OR`ring the metric type.

I'll dig further, but in the meantime, added metrics for remaining and issued tokens, which is closer to what we want here.

File: pkg/controller/issueapi/metrics.go
Comment (sethvargo) [inline pkg/controller/issueapi/metrics.go:124]: Should these be `_count` instead of `_latest`?

Comment (femnad) [reply pkg/controller/issueapi/metrics.go:124]: I thought `_count` was for cases where aggregation is `Count`. But do you mean keep the suffix as `_count` regardless of aggregation type?

Comment (sethvargo) [reply pkg/controller/issueapi/metrics.go:124]: nvm I think you're right

**Comment by sethvargo** [comment]
/lgtm
/hold

I'd like @icco to give the final approval

**Comment by femnad** [comment]
> New changes are detected. LGTM label has been removed.

Had to rebase on main to fix a conflict.

**Review by icco — APPROVED** [review]
/lgtm

**Pull request merged**
